operator/pkg/bgpv2/cluster.go

```diff
@@ -102,31 +119,58 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
 		}
 		nodeSelector = selector
 	}
-
 	// Name of the nodes selected by nodeSelector
 	matchingNodes := sets.New[string]()
-
 	// Name of the ClusterConfig selecting the same node
 	conflictingClusterConfigs := sets.New[string]()
 
-	// Errors
-	var errs error
-
 	for _, node := range b.ciliumNodeStore.List() {
-		if !nodeSelector.Matches(slim_labels.Set(node.Labels)) {
-			continue
-		}
-
-		// Record selected node for later use
-		matchingNodes.Insert(node.Name)
-
-		// Find node config for this node
+		bgpNode := nodeSelector.Matches(slim_labels.Set(node.Labels))
 		oldNodeConfig, oldNodeConfigExists, err := b.nodeConfigStore.GetByKey(resource.Key{Name: node.Name})
+
 		if err != nil {
 			errs = errors.Join(errs, err)
 			continue
 		}
-
+		// restore the router ID from the old NodeConfig if it exists
+		if oldNodeConfigExists && b.bgpRouterIDIPPoolEnabled {
+			for _, instance := range oldNodeConfig.Spec.BGPInstances {
+				if instance.RouterID == nil {
+					continue
+				}
+				routerID, err := netip.ParseAddr(*instance.RouterID)
+				if err != nil {
+					errs = errors.Join(errs, err)
```

Comment:
nit: might be nice to wrap some of these errors, if we end up getting a bunch of joined errors it will be hard to distinguish to what part of this procedure the failure was related to.